jsonschema-core.xml

```diff
@@ -800,49 +800,257 @@
+                        keyword.  As semantic conflicts are not generally detectable through
+                        schema validation, implementations are not expected to detect such
+                        conflicts.  If conflicting vocabularies are declared, the resulting
+                        behavior is undefined.
+                    </t>
+                    <t>
+                        Vocabulary authors SHOULD provide a meta-schema that validates the
+                        expected usage of the vocabulary's keywords.  Such meta-schemas
+                        SHOULD NOT forbid additional keywords, and MUST NOT forbid
+                        the "$id", "$schema", or "$vocabulary" keywords in the root schema.
+                    </t>
+                    <t>
+                        It is RECOMMENDED that meta-schema authors reference each vocabulary's
+                        meta-schema using the <xref target="allOf">"allOf"</xref> keyword,
+                        although other mechanisms for constructing the meta-schema may be
+                        appropriate for certain use cases.
+                    </t>
+                    <t>
+                        Meta-schemas MAY impose additional constraints, including describing
+                        keywords not present in any vocabulary, beyond what the meta-schemas
+                        associated with the declared vocabularies describe.  This allows for
+                        restricting usage to a subset of a vocabulary, and for validating
+                        locally defined keywords not intended for re-use.
+                    </t>
+                    <t>
+                        Meta-schemas intended for local use, with no need to test for
+                        vocabulary support in arbitrary implementations, can safely omit
+                        "$vocabulary" entirely.
+                    </t>
+                </section>
+                <section title="The JSON Schema Core Vocabulary">
+                    <t>
+                        Keywords declared in in this specification that begin with "$" make up
+                        the JSON Schema Core vocabulary.  These keywords are either required in
+                        order process any schema or meta-schema, including those split across
+                        multiple documents, or exist to reserve keywords for purposes that
+                        require guaranteed interoperability.
+                    </t>
+                    <t>
+                        The Core vocabulary MUST be considered mandatory at all times, in order
+                        to bootstrap the processing of further vocabularies.  Meta-schema
+                        authors MAY explicitly list the Core vocabulary, but do not need
+                        to do so.  If present, the Core vocabulary entry MUST have a value
+                        of true indicating that it is required.  The behavior of a value
```

Comment:
Why have a boolean value at all if the value must be `true` and `false` is undefined?  Why not just have an array of URIs?

Reply:
The value of `false` has undefined behavior **for the core vocabulary only**.  The behaviors of the booleans is defined in general earlier in the document.

You can't not use the core vocabulary- you need it to bootstrap processing (assuming you don't just hardcode it, which you can do because it's mandatory and I expect most people will).  Basically, I expect implementations to ignore whether core is present in `$vocabulary` but some people will like having it there.  We could either say that `false` is an error or just say that if you do this and weird stuff happens it's your own fault.  I prefer the latter.

Reply:
@gregsdennis annotations.  Hyper-Schema is a perfect example.  I want to have the hyper-schema meta-schema declare that schemas that `$schema`-link it make use of the core, applicator, assertion, basic annotation, and hyper-schema vocabularies.  The hyper-schema vocabulary is just `links`, `base`, and the sub-keywords inside LDOs.  If you feed a hyper-schema to a validator that has no idea what hyper-schema is, I still want the validator to work.  The hyper-schema keywords do not change standard validation at all.  So one variation on the "hyper-schema" meta-schema will have something like:

```JSON
{
    "https://json-schema.org/draft-08/vocabularies/core": true,
    "https://json-schema.org/draft-08/vocabularies/applicator": true,
    "https://json-schema.org/draft-08/vocabularies/assertions": true,
    "https://json-schema.org/draft-08/vocabularies/basic-annotations": false,
    "https://json-schema.org/draft-08/vocabularies/hyper-schema": false
}
```

Someone might want to use the same meta-schema but with a `$vocabulary` where all four of those are `true`, because they know they are only using it where hyper-schema and annotations like `readOnly` are critical.

So we can write a meta-schema that validates all of these, and then publish a meta-schema that references it with this set of vocabularies (mandatory validation, optional annotations and hyper-schema), and one that makes them all mandatory.  Different options will be preferred for different use cases.

Reply:
So if my implementation doesn't know what `basic-annotations` is, it just leaves out those annotations?  This means that implementations will need to expose a way to register new vocabularies by their spec (e.g. `https://json-schema.org/draft-08/vocabularies/basic-annotations`) that then links them to their associated meta-schema (e.g. `https://json-schema.org/draft-08/basic-annotations`).  Only then will my implementation know what to do with that.

Also, the `false` allows my implementation to continue if it doesn't know about `basic-annotations`, but it should return an error if it doesn't know about `assertions`.

Is all that right?